Fix aircraft reservation madness

obrakmann opened the pull request:
Fixes #11953.
Fixes #5995.
Fixes #7545.
Fixes #7481.

The important part of this change is the `MayYieldReservation` flag on `Aircraft`. It is set after an aircraft finished resupplying, and will cause it to be kicked off the airfield by the `Reservable` trait if another aircraft tries to get a reservation.

One change wrt to bleed that I couldn't find a solution for is that producing a plane will block the newly-built plane as long as the other one is still resupplying, but will cause both to take off once its done.

Even with this change, there can still (rarely) arise situations where airfields will get locked by a reservation. From the way it looked in my tests, it seemed that the aircraft that took the reservation was unable to find the proper approach to land on the airfield. I didn't yet look deeper into it, as it seemed unrelated to the issue being dealt with by this PR.

obrakmann committed:
Fix aircraft reservation madness

obrakmann commented:
As per discussion on IRC, added a commit that disables auto-takeoff for aircraft.

pchote commented on OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs:
```diff
@@ -27,22 +28,29 @@ public ResupplyAircraft(Actor self)
-			// If is helicopter move away as soon as the resupply ends
-			return ActivityUtils.SequenceActivities(
-				aircraft.GetResupplyActivities(host).Append(new TakeOff(self)).Append(NextActivity).ToArray());
+			if (inner == null)
```
IMO this would be clearer as

```csharp
if (inner != null)
{
	inner = ActivityUtils.RunActivity(self, inner);
	return this;
}
// your if case with one less level of indentation
```

obrakmann replied:
I don't like that for two reasons:

1. It makes the method harder to read. You'd have the stuff that happens in subsequent ticks at the top, and the thing that runs during the first tick at the bottom.
2. I'd need two return statements.

pchote replied:
Ok.

pchote commented on OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs:
```diff
@@ -27,22 +28,29 @@ public ResupplyAircraft(Actor self)
+				if (host == null)
+					return NextActivity;
+
+				inner = aircraft.IsPlane ?
```
This feels like an abuse of the terniary statement.  Can we have an if-else?

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Reservable.cs:
```diff
@@ -29,11 +29,18 @@ public void Tick(Actor self)
+			{
+				/* not likely to arrive now. */
+				reservedFor = null;
+				reservedForAircraft = null;
```
I wonder if not calling `reservedForAircraft.UnReserve();` here might cause your reservation issue.

pchote commented:
Can we please have the `Production` trait try to boot off a parked plane if there are no free spots?

pchote commented on OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs:
```diff
@@ -27,22 +28,29 @@ public ResupplyAircraft(Actor self)
+
+				inner = aircraft.IsPlane ?
+					ActivityUtils.SequenceActivities(
+						aircraft.GetResupplyActivities(host)
```
Shifting this to an inner activity seems to have regressed the ability to cancel mid-reload and fly away to do something else.

pchote commented:
If you have multiple circling empty aircraft then issue the stop command they will all land on the same empty airfield to reload.

obrakmann commented:
> Can we please have the Production trait try to boot off a parked plane if there are no free spots?

It already does? Unless all aircraft have just been built and never did anything else, then it'll block the last produced plane. But once they've been airborne, they'll always be in the `ResupplyAircraft` activity, which eventually sets the flag that allows the plane to get booted off.

obrakmann commented:
> If you have multiple circling empty aircraft then issue the stop command they will all land on the same empty airfield to reload.

Can't reproduce, works fine here. What did you do exactly?

I broke helicopter-reloading, though.

pchote commented:
Here are my steps:

1. Start game on ra mod, a51d9e5688fae6b06e7ebadd5e6c290749dd7fdc
2. `/all` to enable cheats.
3. Build airfield then two migs (one second apart).  Observe that first mig does not take off the make room for the second.
4. Order it to take off, then drag select both migs.  Tell them to shoot stuff until they are half ammo, then move to a location and circle.
5.  Select both migs then press `s`.  Observe them both return to base and land on the same airfield.

pchote commented:
> Unless all aircraft have just been built and never did anything else, then it'll block the last produced plane.

Ah, that is what i'm encountering.  If we can't fix that we should restore the rallypoint move on production.

obrakmann commented:
> Here are my steps:

Interesting. I get exactly the opposite result: the airfield stays locked and both migs start circling. That seems to be another bug. It again only happens to newly-produced units. If you order it to attack something right from the airfield, it will not un-reserve. If you order it to move, however, it will.

obrakmann committed:
Fix FlyAttack not calling TakeOff for landed aircraft

obrakmann commented:
Updated.
* Addressed review comments.
* Removed the commit that disabled the auto-takeoff for planes again.
* Un-broke helicopter-reloading (the activity queue would get lost on the first iteration of this PR).

pchote commented:
I'm not sure if it's a regression from this PR or an earlier one, but if you target something a plane will use all its ammo, reload, and then go to the rallypoint instead of continuing its attack.

obrakmann committed:
Make aircraft continue attack on same target after re-arming

pchote commented:
Confirmed that planes are now working properly, thanks.

Helicopters will need a similar fix.

pchote commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -56,6 +57,10 @@ public override Activity Tick(Actor self)
 				else
 					inner = ActivityUtils.SequenceActivities(new Fly(self, target), new FlyTimed(ticksUntilTurn, self));
+
+				// This needs to be done in this round-about way because TakeOff doesn't behave as expected when it doesn't have a NextActivity.
```
Can we prepend a `// HACK: ` to this?  We really should fix TakeOff in a future PR.

pchote commented on OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs:
```diff
@@ -27,22 +28,45 @@ public ResupplyAircraft(Actor self)
+				else
+				{
+					// Helicopters should take off from their helipad immediately after resupplying.
+					// NextActivity needs to be appended here because otherwise TakeOff does stupid things.
```
Could do with `// HACK: `s here too.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Reservable.cs:
```diff
@@ -29,11 +29,19 @@ public void Tick(Actor self)
 			if (!Target.FromActor(reservedFor).IsValidFor(self))
-				reservedFor = null;		/* not likely to arrive now. */
+			{
+				/* not likely to arrive now. */
```
Minor style nit: `// Not likely to arrive now` please.

pchote commented:
All looks good to me now, great work :+1:

Mailaender commented:
This does work better than before in my stress tests. :white_check_mark: However I don't understand most of the code interactions.

pchote commented:
Just a note that i've added #5995, #7545, #7481 to the list of issues fixed in the main description.
